File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root || p->val==root->val||q->val==root->val) return root;
        int pv = p->val,qv=q->val,v=root->val;
        if(pv>v && qv>v){
            return lowestCommonAncestor(root->right, p,q);
        }
        else if(pv<v && qv<v){
            return lowestCommonAncestor(root->left, p,q);
        }
        else{
            return root;
        }
    }
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        temp = root
        smaller = p if p.val<q.val else q
        larger = p if smaller==q else q
        while temp:
            if temp==smaller or temp==larger:
                return temp
            if smaller.val<temp.val<larger.val:
                return temp
            if larger.val<temp.val:
                temp = temp.left
            else:
                temp = temp.right
        return temp
            